Add Vote Account Management Doc (#11278)
solana-labs/solana: docs/src/running-validator/vote-accounts.md

---
title: Vote Account Structure and Configuration
---

This page describes how to set up an on-chain vote account.   Creating a vote
account is needed if you plan to run a validator node on Solana.

All settings for a vote account can be configured at the time the account is
created, or they can be set after the account is created while the corresponding
validator is running.

### Vote Account Address

A vote account is created at an address that is either the public key of a
keypair file, or at a derived address from the authorized voter's public key.

The address of a vote account does not need to have a stored corresponding
private key, because the address (although it may be a valid public key) is never
needed to sign any transactions, but is just used to look up the account
information.

When someone wants to [delegate tokens in a stake account](../staking.md),
the delegation command is pointed at the vote account address of the validator
to whom the token-holder wants to delegate.

### Validator Identity

> The Validator Identity is a system account that pays for all the vote transaction

Reviewer: Use italics for new terms. And only capitalize proper nouns

fees submitted by the vote account.  Because the vote account is expected to vote
on most valid blocks it receives, the Validator Identity account is frequently
(as often as once per slot, or roughly every 400ms) signing transactions and
paying fees.  For this reason the Validator Identity account must be stored as
a "hot wallet" as a keypair file on the same system the validator process is
running.

Because a hot wallet is generally less secure than an offline or "cold" wallet,
we only recommend storing enough SOL on the Identity keypair to cover voting fees
for some amount of time (perhaps a few weeks or months) and periodically topping
off your Validator Identity account from a more secure wallet.

The Validator Identity is required to be provided when a vote account is created.
The Validator Identity can also be changed after an account is created by using
the `vote-update-validator` command.

> ### Authorized Voter

Reviewer: To be consistent with other authority terminology, this should be called "vote authority"

Identity.  If the Validator Identity is also the Authorized Voter, only one
signature per vote transaction is needed in order to both sign the vote and pay
the transaction fee.  Because transaction fees on Solana are assessed
per-signature, having one signer instead of two will result in half the transaction
fee paid compared to setting the Voter and Identity to two different accounts.

The Authorized Voter can be set when the vote account is created.  If it is not
provided, the vote account with set the Validator Identity as the Authorized
Voter by default.  The Authorized Voter can be changed later with the
`vote-authorize-voter` command.

The Authorized Voter can be changed at most once per epoch.  If the Voter is
changed with `vote-authorize-voter`, this will not take effect until the end of
the current epoch.  To support a smooth transition of the vote signing,
`solana-validator` allows the `--authorized-voter` argument to be specified
multiple times.  This allows the validator process to keep voting successfully
when the network reaches an epoch boundary at with the validator's authorized
voter account changes.

> ### Authorized Withdrawer

Reviewer: Likewise, withdraw authority

---
title: Vote Account Management
---

This page describes how to set up an on-chain _vote account_.   Creating a vote
account is needed if you plan to run a validator node on Solana.

All settings for a vote account can be configured at the time the account is
created, or they can be set after the account is created while the corresponding
validator is running.

### Vote Account Address

A vote account is created at an address that is either the public key of a

> keypair file, or at a derived address from the authorized voter's public key.

Reviewer 2: A derived address doesn't technically have to be based off either authority. It can be any ole keypair.  Perhaps that technicality doesn't need communicated here though?

Author: I can leave the technicality out, but the CLI description is actually rather confusing to me, looking at it again:
```
        --seed <STRING>
            Seed for address generation; if specified, the resulting account will be at a derived address of the VOTE
            ACCOUNT pubkey
```
That's the only place we use the term VOTE ACCOUNT in the help text.  I assume this is referring to the vote account keypair positional arg?  In which case my description here about deriving from the vote auth pubkey is just wrong anyway.

> keypair file, or at a derived address from the authorized voter's public key.

Reviewer: Authorized voter

The address of a vote account does not need to have a stored corresponding
private key, because the address (although it may be a valid public key) is never
needed to sign any transactions, but is just used to look up the account
information.

When someone wants to [delegate tokens in a stake account](../staking.md),
the delegation command is pointed at the vote account address of the validator
to whom the token-holder wants to delegate.

### Validator Identity

> The _validator identity_ is a system account that pays for all the vote transaction

Reviewer: An account doesn't pay. It's used to pay

fees submitted by the vote account.  Because the vote account is expected to vote
on most valid blocks it receives, the validator identity account is frequently

> (as often as once per slot, or roughly every 400ms) signing transactions and

Reviewer 2: ```suggestion
(potentially multiple times per second) signing transactions and
```

One fewer place to update when toly gets his 200ms slots :wink:

> paying fees.  For this reason the validator identity account must be stored as

Reviewer: The account isn't stored in a wallet, it's keypair is

> a "hot wallet" as a keypair file on the same system the validator process is

Reviewer 2: ```suggestion
a "hot wallet" in a keypair file on the same system the validator process is
```

running.

Because a hot wallet is generally less secure than an offline or "cold" wallet,

> we only recommend storing enough SOL on the Identity keypair to cover voting fees

Reviewer: You store SOL in an account, not on a keypair

for some amount of time (perhaps a few weeks or months) and periodically topping
off your validator identity account from a more secure wallet.

The validator identity is required to be provided when a vote account is created.
The validator identity can also be changed after an account is created by using
the `vote-update-validator` command.

### Vote Authority

The _vote authority_ keypair is used to sign each vote transaction the validator

> node wants to submit to the cluster.  This can be the same keypair used as the
> validator identity for the same vote account, or a different keypair.  Because

Reviewer 2: ```suggestion
node wants to submit to the cluster.  This doesn't necessarily have to be unique
from the validator identity, as you will see later in this document.  Because
```

the vote authority, like the validator identity, is signing transactions
frequently, this also must be a hot keypair on the same file system as the
validator process.

It is recommended to set the vote authority to the same address as the validator
identity.  If the validator identity is also the vote authority, only one
signature per vote transaction is needed in order to both sign the vote and pay
the transaction fee.  Because transaction fees on Solana are assessed
per-signature, having one signer instead of two will result in half the transaction
fee paid compared to setting the vote authority and validator identity to two
different accounts.

The vote authority can be set when the vote account is created.  If it is not

> provided, the vote account with set the Validator Identity as the vote authority

Reviewer: Caps

> provided, the vote account with set the Validator Identity as the vote authority
> by default.  The vote authority can be changed later with the

Reviewer 2: ```suggestion
provided, the default behavior is to assign it the same as the Validator Identity.
The vote authority can be changed later with the
```

`vote-authorize-voter` command.

The vote authority can be changed at most once per epoch.  If the authority is

> changed with `vote-authorize-voter`, this will not take effect until the end of
> the current epoch.  To support a smooth transition of the vote signing,

Reviewer 2: ```suggestion
changed with `vote-authorize-voter`, this will not take effect until the beginning
of the next epoch.  To support a smooth transition of the vote signing,
```

`solana-validator` allows the `--authorized-voter` argument to be specified
multiple times.  This allows the validator process to keep voting successfully
when the network reaches an epoch boundary at with the validator's vote authority
account changes.

### Withdraw Authority

The _withdraw authority_ keypair is used to withdraw funds from a vote account
using the `withdraw-from-vote-account` command.  Any network rewards a validator
earns are deposited into the vote account and are only retrievable by signing
with the withdraw authority keypair.

The withdraw authority is also required to sign any transaction to change

> a vote account's [commission](#commission).

Reviewer 2: I believe it is required to sign for _all_ account settings modifying transactions

Author: It is needed to change the withdrawer (clearly) and also to change the validator identity, but not needed to set a new vote authority.

`--authorized-withdrawer` option.  If this is not provided, the validator
identity will be set as the withdraw authority by default.

The withdraw authority can be changed later with the `vote-authorize-withdrawer`
command.

### Commission

_Commission_ is the percent of network rewards earned by a validator that are
deposited into the validator's vote account.  The remainder of the rewards
are distributed to all of the stake accounts delegated to that vote account,
proportional to the active stake weight of each stake account.

For example, if a vote account has a commission of 10%, for all rewards earned
by that validator in a given epoch, 10% of these rewards will be deposited into
the vote account in the first block of the following epoch. The remaining 90%
will be deposited into delegated stake accounts as immediately active stake.

Commission can be set upon vote account creation with the `--commission` option.
If it is not provided, it will default to 100%.

Commission can also be changed later with the `vote-update-commission` command.

When setting the commission, only integer values in the set [0-100] are accepted.

> The integer represent the number of percentage points for the commission, so

Reviewer 2: ```suggestion
The integer represents the number of percentage points for the commission, so
```

---
title: Vote Account Management
---

This page describes how to set up an on-chain _vote account_.   Creating a vote
account is needed if you plan to run a validator node on Solana.

> All settings for a vote account can be configured at the time the account is

Reviewer: Not sure I grok this sentence. It sounds like it could be reduced to to "The vote account can be configured when first created or after the validator is running."  Also, should there be some details on how to do that or what the configuration options are?

keypair file, or at a derived address based on a keypair file's public key and
a seed string.

The address of a vote account does not need to have a stored corresponding
private key, because the address (although it may be a valid public key) is never
needed to sign any transactions, but is just used to look up the account
information.

When someone wants to [delegate tokens in a stake account](../staking.md),
the delegation command is pointed at the vote account address of the validator
to whom the token-holder wants to delegate.

### Validator Identity

The _validator identity_ is a system account that is used to pay for all the
vote transaction fees submitted by the vote account.
Because the vote account is expected to vote on most valid blocks it receives,
the validator identity account is frequently
(potentially multiple times per second) signing transactions and
paying fees.  For this reason the validator identity account's keypair must be
stored as a "hot wallet" in a keypair file on the same system the validator
process is running.

Because a hot wallet is generally less secure than an offline or "cold" wallet,

> we only recommend storing enough SOL on the identity account to cover voting fees

Reviewer: Avoid making recommendations. Instead, state either the benefits of taking some action and/or the consequences of not taking it.

for some amount of time (perhaps a few weeks or months) and periodically topping
off your validator identity account from a more secure wallet.

The validator identity is required to be provided when a vote account is created.
The validator identity can also be changed after an account is created by using
the `vote-update-validator` command.

### Vote Authority

The _vote authority_ keypair is used to sign each vote transaction the validator
node wants to submit to the cluster.  This doesn't necessarily have to be unique
from the validator identity, as you will see later in this document.  Because
the vote authority, like the validator identity, is signing transactions
frequently, this also must be a hot keypair on the same file system as the
validator process.

> It is recommended to set the vote authority to the same address as the validator

Reviewer: Another recommendation, but a little worse, because it's in the passive voice. Who recommends?

identity.  If the validator identity is also the vote authority, only one
signature per vote transaction is needed in order to both sign the vote and pay
the transaction fee.  Because transaction fees on Solana are assessed
per-signature, having one signer instead of two will result in half the transaction
fee paid compared to setting the vote authority and validator identity to two
different accounts.

The vote authority can be set when the vote account is created.  If it is not
provided, the default behavior is to assign it the same as the validator identity.
The vote authority can be changed later with the `vote-authorize-voter` command.

The vote authority can be changed at most once per epoch.  If the authority is
changed with `vote-authorize-voter`, this will not take effect until the
beginning of the next epoch.  To support a smooth transition of the vote signing,
`solana-validator` allows the `--authorized-voter` argument to be specified
multiple times.  This allows the validator process to keep voting successfully

> when the network reaches an epoch boundary at with the validator's vote authority

Reviewer: "at with"?

account changes.

### Withdraw Authority

The _withdraw authority_ keypair is used to withdraw funds from a vote account
using the `withdraw-from-vote-account` command.  Any network rewards a validator
earns are deposited into the vote account and are only retrievable by signing
with the withdraw authority keypair.

The withdraw authority is also required to sign any transaction to change
a vote account's [commission](#commission), and to change the validator
identity on a vote account.

> Because the vote account could accrue a significant balance, it is recommended

Reviewer: Instead of "it is recommended to keep", consider "consider keeping"

to keep the withdraw authority keypair in an offline/cold wallet, as it is
not needed to sign frequent transactions.

The withdraw authority can be set at vote account creation with the
`--authorized-withdrawer` option.  If this is not provided, the validator
identity will be set as the withdraw authority by default.

The withdraw authority can be changed later with the `vote-authorize-withdrawer`
command.

### Commission

_Commission_ is the percent of network rewards earned by a validator that are
deposited into the validator's vote account.  The remainder of the rewards
are distributed to all of the stake accounts delegated to that vote account,

> proportional to the active stake weight of each stake account.

Reviewer: How about another sentence or two explaining why a validator might set a commission to 0%, 50% or 100%?
